### PR TITLE
feat(ledger): persist genesis trie nodes for L2 chains (M6 follow-up)

### PR DESCRIPTION
**Pull request by kyonRay**
## Failing scenario

An L2 (scenario-B) chain with non-empty genesis allocs: `Ledger::buildGenesisBlock` computes the op-geth-compatible state root over the allocs (`computeGenesisStateRoot`, Ledger.cpp) and seals it into the genesis header — but only the root. The produced trie nodes are discarded (`GenesisStateRoot.cpp` said so explicitly: *"the produced nodes are not persisted — only the root goes into the genesis header"*). Block 1's execute-time incremental MPT build (#5345) then takes the genesis root as parent and must read the parent trie's nodes; none exist, so block 1 fails with `MPTInvariantViolation: Trie: missing node hash`. This was a declared scope cut in #5345; this PR closes it.

## Fix

- `GenesisStateRoot.h/.cpp`: new `GenesisStateTrie{root, nodes}` + `computeGenesisStateTrie()`. `computeTrieRoot` already returns `TrieBuildResult{root, newNodes}`, so no trie API changes — the account trie's nodes and every per-account storage sub-trie's nodes are merged into one hash-keyed map. `computeGenesisStateRoot` keeps its signature and delegates.
- `Ledger.cpp` (`buildGenesisBlock`, fresh-init path only): right after `importGenesisState`, write every node as an ordinary state row `StateKey{"/mpt/", digest}` via `storage2::writeOne(*m_stateStorage, mptNodeStateKey(h), Entry(rlp))` — the identical write path the alloc flat rows use, so node rows share the genesis batch's fate. Gated on `feature_l2_ethereum_compat` in the genesis features (the same predicate as `NodeConfig::validateL2Invariants` and `shouldBuildMPT`'s scenario-B branch).

Behavior notes: the exists-genesis restart path returns before the write loop, so restarts never re-execute it. Scenario A (upgrade activation) is unaffected — its first MPT block starts from `emptyRootHash()` and needs no genesis nodes. Non-MPT chains and allocs-without-flag chains write zero rows.

Build note: linking the `storage` target into bcos-ledger breaks the `install(EXPORT fiscobcosTargets)` set (storage is not exported), so the dependency is an include-dir only (`KeyPrefixes.h` is header-only), PRIVATE and build-tree only.

## Tests

New `test_GenesisNodePersistence.cpp`, 4 cases:

- (a) walk the trie from the genesis header stateRoot via NodeDecoder over the stored rows, checking each node's content hash and descending into storage sub-tries via the decoded account's storageRoot; also asserts stored-row-count == visited-set (no garbage rows);
- (b) empty allocs → zero node rows;
- (c) pbft chain and allocs-without-flag → zero node rows;
- (d) end-to-end: `buildAndCollect` with the genesis root as parent and a block-1-style delta (balance update + new storage slot, subsequent-touch) succeeds, and the resulting root equals an independent from-scratch `computeTrieRoot` oracle.

Red before the fix — cases (a) and (d) fail, (d) with the exact `MPTInvariantViolation: Trie: missing node hash` from the failing scenario. Green after. Negative control: disabling the gate re-produces both failures. Full `test-bcos-ledger` and `test-transaction-scheduler` (comment-only change there) both pass with `*** No errors detected`.

## Known limitation (pre-existing, not introduced here)

A hypothetical empty-alloc L2 genesis would leave the header stateRoot as zero-h256 while `commitTrie` treats only `emptyRootHash()` as from-empty, so block 1 would still fail on such a chain. Currently unreachable: `validateL2Invariants` requires non-empty allocs together with the flag. Left for a follow-up note rather than widening this PR.

## Appended: end-to-end integration tests (M6.4a + M6.5 + M8.4)

Three test-only commits stacked on top (zero production-code changes):

- `d4179c3aa` — `FullChainFixture`: a reusable end-to-end harness over a real `CheckpointRocksDBStorage` (the production four-parameter type), real `Ledger::buildGenesisBlock` (so this PR's genesis node persistence is exercised naturally) and the real `BaselineScheduler` execute/commit path including `prewriteBlockToBuffer`. Feature activation is driven by writing real SYS_CONFIG rows (`{value="1", enableNumber=N}`, the same shape governance writes) read back through `Features::readFromStorage` — no test-only injection hooks.
- `8fa21a816` — activation-transition tests: scenario A (block N still XOR → N+1 switches to MPT with an empty-trie parent, matching a from-scratch oracle → N+2 continues incrementally; dormant-account first-touch commits only touched rows, pinned by a negative oracle) and scenario B (genesis root equals an independent RLP+`computeTrieRoot` oracle; block 1 builds incrementally on the persisted genesis nodes — the first full-stack verification of this PR's fix; L2 takes priority over a scenario-A activation setting).
- `2c8c9572a` — eth_getProof end-to-end: active account → proof verifies; dormant scenario-A account → `-32004` per `EthEndpoint.cpp`; scenario B allocs all prove and verify; a proof checked against a wrong root must be rejected.

Both `test-transaction-scheduler` and `test-bcos-rpc` full suites pass. One note for future work: linking `executor` into `test-bcos-rpc` breaks exception catching binary-wide (its PUBLIC `wedprcrypto` Rust static library ships a conflicting runtime — a libc++ `std::invalid_argument` thrown by `stoull` becomes uncatchable). The tests therefore provide a linkage-only copy of `unhexAddress` instead of linking `executor`; a future duplicate symbol would fail loudly at link time.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: bcos-ledger/bcos-ledger/Ledger.cpp [hunk 1]
```diff
@@ -2142,6 +2146,29 @@ bool Ledger::buildGenesisBlock(
...
+        // account's storage sub-trie — must be persisted here, on the same storage the alloc
+        // flat rows above just went to. A missing node aborts block-1 execution loudly
+        // (MPTInvariantViolation). Non-MPT chains never read these rows and get none; scenario
+        // A (feature_mpt_state_root activated mid-chain) starts its first MPT block from
+        // emptyRootHash() and needs no genesis nodes either.
+        if (std::any_of(genesis.m_features.begin(), genesis.m_features.end(),
+                [](ledger::FeatureSet const& featureSet) {
+                    return featureSet.flag == Features::Flag::feature_l2_ethereum_compat &&
+                           featureSet.enable > 0;
+                }))
```
Comment (ywy2090) [inline bcos-ledger/bcos-ledger/Ledger.cpp:2161]: 💡 [SUGGESTION] 这个谓词与 `NodeConfig::validateL2Invariants`(`bcos-tool/bcos-tool/NodeConfig.cpp:301-304`)逐字重复——同一份数据源(`genesis.m_features`)、同一个 lambda。

两处失步的后果不对称:若这里判成 false 而校验那边判成 true,创世节点不会被写,而链能正常启动,直到块 1 才报 `MPTInvariantViolation` ——正是本 PR 要消灭的那个故障,且此时创世已封、无法回填(重启走 `:1976` 早返回)。

建议把它提到 `GenesisConfig.h` 或 `Features.h` 上做成一个共享的 `isL2GenesisEnabled(GenesisConfig const&)`,两边都调它。这样「L2 链的判定标准」只有一处定义,也顺带让 `enable` 在创世配置里是布尔(而非激活块号)这个反直觉的语义只需要解释一次。

**Review by ywy2090 — APPROVED** [review]
## 概述

本 PR 补上 #5345 声明的范围裁剪:L2(scenario B)链的创世状态根此前只把根塞进创世头、**丢弃产生的 trie 节点**,于是块 1 的增量 MPT 构建以创世根为父节点去读父 trie,一个节点都读不到,报 `MPTInvariantViolation: Trie: missing node hash`。改法是新增 `computeGenesisStateTrie()` 一并返回根与全部节点(账户 trie + 每个账户的存储子 trie),并在 `buildGenesisBlock` 的全新初始化路径上把它们作为普通 `/mpt/` 状态行写入。7 个文件 +520/-20。

**实现正确。** 这是创世期的不可逆写入,我按此标准逐条核实了边界条件,未发现正确性问题。

## 逐条核实的论断

- **门控谓词正确 —— 这一条我一开始判断错了,核实后推翻。** 看到 `featureSet.enable > 0` 时我怀疑是反的:scenario B 要求该 flag 在**创世(block 0)**激活,若 `enable` 是激活块号,`enable > 0` 恰好表示「块 1 及以后激活」。查证后发现创世配置的解析路径是 `it.second.get_value<bool>()` 再转 int(`NodeConfig.cpp:1544-1547`),即 `enable` 在创世配置里是 **0/1 布尔**,不是激活块号。因此 `enable > 0` 就是「已启用」,且与 `validateL2Invariants`(`NodeConfig.cpp:301-304`)的 lambda 逐字一致。
- **重启路径确实不会重复执行。** `existsGenesisData == genesisData` 分支在 `Ledger.cpp:1976` 就 `co_return true`,远早于 `:2149` 的写循环;创世数据不一致的分支直接抛异常。写循环只在全新初始化时可达。
- **从空构建 ⇒ `newNodes` 即全部节点。** 账户 trie 与每个存储子 trie 都是 from-empty 走无状态核心,所以增量语义下的「新节点」等于「全部节点」。测试用例 (a) 从创世头的 stateRoot 出发经 `NodeDecoder` 实走整棵树、逐节点校验内容哈希并下潜存储子 trie,直接验证了这个假设。
- **`nodes.merge()` 对内容寻址节点安全。** `unordered_map::merge` 遇到重复 key 不移动、保留原值;而节点按内容哈希索引,重复 key 意味着内容相同,保留哪个都一样。
- **移动语义安全。** 写循环用 `std::move(nodeRlp)` 掏空 map 的值,但 `ethStateTrie.root`(h256,值语义)早在 `:2069` 的 `setStateRoot` 用过,之后不再读 `nodes`。
- **已知限制确实不可达。** 空 alloc 的 L2 创世会让头 stateRoot 保持零值、而 `commitTrie` 只把 `emptyRootHash()` 当作 from-empty ——但 `validateL2Invariants` 明确拒绝「启用 flag 且 allocs 为空」的配置,故当前无法构造。PR 如实记录了这一点而不是顺手扩大改动范围,这个处理是对的。
- **原子性符合描述。** 节点行经 `storage2::writeOne(*m_stateStorage, ...)` 写入,与上方 `importGenesisState` 写 alloc 扁平行的是同一个 storage、同一次创世提交(`Ledger.cpp:2300` 注释亦述及),因此「node rows share the genesis batch's fate」成立。
- **CMake 取巧方案成立。** `KeyPrefixes.h` 只 include `StateKey.h`、`FixedBytes.h`、`<string_view>`,全部来自已链接的 bcos-framework / bcos-utilities,确为 header-only;只加 PRIVATE include 目录、不链接 `storage`,避免把 RocksDB 拖进导出的 ledger 目标。

## 测试

四个用例分工清晰,且都有判别力:(a) 实走整棵 trie 校验每个节点的内容哈希,并断言**存储行数 == 访问集合**(既证完整、又证无垃圾行);(b)(c) 空 allocs / pbft 链 / 有 allocs 无 flag 三种否定情形均写零行;(d) 端到端以创世根为父跑 `buildAndCollect`,并与独立的 from-scratch `computeTrieRoot` 预言机比对根值。PR 还给出了红/绿证据与关闭门控的反向对照,(d) 复现的正是失败场景里那条 `MPTInvariantViolation`。这比「测试通过」有说服力得多。

## 一条运维提示(非缺陷)

补写只发生在全新初始化路径,所以**在本 PR 之前用同一开发分支建过的 L2 测试链无法被回填**:重启走 `:1976` 早返回,节点行永远不会补上,块 1 依旧失败。

我查证了这个问题的影响范围:`feature_l2_ethereum_compat` **未出现在任何已发布 tag 中**,整条 L2/MPT 路径都还在 release-3.18.0 开发分支上,所以不涉及任何生产链,只影响开发者手里的旧测试链——处理方式是清库重建。不作为缺陷,但值得在合并说明里提一句,能省掉别人一轮困惑的排查。

## 建议

创世 L2 判定谓词现在有两处逐字重复(`NodeConfig.cpp:303` 与本 PR 的 `Ledger.cpp:2159`),见内联建议。

## CI

三个必需 Build(macos-15 / ubuntu-24.04 / ubuntu-24.04-arm)+ Coverage + Codacy + Hound 全绿。`macos-15-intel` 与 `windows-2025` 的失败为 pre-existing,与本 diff 无关。

## 结论

Found: 0 critical, 0 important, 1 suggestion

未发现正确性、安全或数据完整性问题。门控谓词(我最初的怀疑经查证被推翻)、重启幂等性、from-empty 的节点完备性、移动语义、写入原子性、CMake 依赖边界均已核实成立;测试有判别力并附红绿证据;已知限制经查证确不可达且如实记录。唯一一条建议是谓词去重,不阻塞合入。

**Approve**

File: bcos-ledger/CMakeLists.txt [hunk 2]
```diff
@@ -25,6 +25,10 @@ target_include_directories(ledger PUBLIC
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
     $<INSTALL_INTERFACE:include/bcos-ledger>)
 target_link_libraries(ledger PUBLIC tool protocol-tars codec table bcos-concepts bcos-crypto Boost::serialization)
+# bcos-storage/KeyPrefixes.h (the "/mpt/" node-row key layout) is header-only; a full
+# `storage` link would pull RocksDB into the exported ledger target, so only the include
```
Comment (morebtcg) [inline bcos-ledger/CMakeLists.txt:29]: **⚠️ 构建可维护性**: 通过 `target_include_directories(ledger PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/../bcos-storage)` 绕过链接依赖是务实的，但 `../bcos-storage` 相对路径依赖于源码树布局。建议在顶层 `CMakeLists.txt` 中为 `bcos-storage` 定义一个 `INTERFACE` 库目标（仅包含头文件路径），让 `ledger` 通过 `target_link_libraries(ledger PRIVATE bcos-storage-headers)` 引用。这样当目录结构变化时只需改一处。

File: bcos-ledger/bcos-ledger/GenesisStateRoot.h [hunk 3]
```diff
@@ -46,4 +48,19 @@ namespace bcos::ledger
 // go-ethereum vectors), so given correct per-account encoding the returned root
 // is op-geth-compatible by construction.
 bcos::task::Task<bcos::h256> computeGenesisStateRoot(GenesisConfig const& genesis);
+
+/// The genesis state trie in full: the op-geth-compatible root plus EVERY node the build
+/// produced — account trie and each account's storage sub-trie — as hash-keyed raw RLP.
+/// Identical encodings across sub-tries hash identically and dedupe in the map.
+struct GenesisStateTrie
+{
+    bcos::h256 root;
```
Comment (morebtcg) [inline bcos-ledger/bcos-ledger/GenesisStateRoot.h:57]: **💡 可扩展性**: `GenesisStateTrie::nodes` 使用 `std::unordered_map<h256, bytes>` 是正确的——跨子 trie 的相同哈希自动去重。但需要注意：`std::unordered_map` 的迭代顺序是不确定的（依赖 hash 函数和 bucket 布局），这意味着 Ledger.cpp 中的写入循环在不同编译/运行环境下可能以不同顺序写入节点行。虽然不影响正确性（每个 key 唯一），但若未来有基于写入顺序的假设（如增量备份），可能会产生意外行为。文档注释中已提到 dedup，建议补充说明迭代顺序不保证。

File: bcos-ledger/bcos-ledger/GenesisStateRoot.h [hunk 4]
```diff
@@ -46,4 +48,19 @@ namespace bcos::ledger
 // go-ethereum vectors), so given correct per-account encoding the returned root
 // is op-geth-compatible by construction.
 bcos::task::Task<bcos::h256> computeGenesisStateRoot(GenesisConfig const& genesis);
+
+/// The genesis state trie in full: the op-geth-compatible root plus EVERY node the build
+/// produced — account trie and each account's storage sub-trie — as hash-keyed raw RLP.
+/// Identical encodings across sub-tries hash identically and dedupe in the map.
```
Comment (morebtcg) [inline bcos-ledger/bcos-ledger/GenesisStateRoot.h:54]: **💡 可扩展性**: `GenesisStateTrie` 是一个纯数据结构，建议考虑将其放入独立的轻量头文件中（如 `GenesisStateTrie.h`），与 `computeGenesisStateTrie` 声明分离。当前放在 `GenesisStateRoot.h` 中意味着任何只需要这个 struct 的消费者都会拉入 `Task.h` 协程依赖。不过在当前只有一个消费者 (Ledger.cpp) 的场景下可以接受。

File: bcos-ledger/bcos-ledger/GenesisStateRoot.cpp [hunk 5]
```diff
@@ -59,21 +59,30 @@ bcos::task::Task<bcos::h256> storageRootOf(std::vector<Alloc::State> const& stor
...
+{
+    auto trie = co_await computeGenesisStateTrie(genesis);
+    co_return trie.root;
+}
+
+bcos::task::Task<bcos::ledger::GenesisStateTrie> bcos::ledger::computeGenesisStateTrie(
+    GenesisConfig const& genesis)
 {
     std::map<bcos::h256, bcos::bytes> stateEntries;
+    std::unordered_map<bcos::h256, bcos::bytes> nodes;
```
Comment (morebtcg) [inline bcos-ledger/bcos-ledger/GenesisStateRoot.cpp:79]: **代码风格**: `nodes.merge(storageTrie.newNodes)` 是对 `std::unordered_map::merge` 的恰当使用——C++17 语义保证了相同 key 时保留已有元素（first-wins），这对于 trie 节点去重是正确的：相同的 RLP 编码产生相同的哈希，保留第一个即可。建议在此行上方加一个简短注释说明 merge 的 first-wins 语义是有意为之。

File: bcos-ledger/bcos-ledger/Ledger.cpp [hunk 6]
```diff
@@ -2142,6 +2146,29 @@ bool Ledger::buildGenesisBlock(
...
+        // Scenario B (L2): block 1 builds the MPT incrementally on top of the genesis state
+        // root (buildAndCollect with the genesis root as parent) and reads the parent trie
+        // through "/mpt/" state rows, so every genesis trie node — account trie plus each
+        // account's storage sub-trie — must be persisted here, on the same storage the alloc
+        // flat rows above just went to. A missing node aborts block-1 execution loudly
+        // (MPTInvariantViolation). Non-MPT chains never read these rows and get none; scenario
+        // A (feature_mpt_state_root activated mid-chain) starts its first MPT block from
+        // emptyRootHash() and needs no genesis nodes either.
+        if (std::any_of(genesis.m_features.begin(), genesis.m_features.end(),
+                [](ledger::FeatureSet const& featureSet) {
```
Comment (morebtcg) [inline bcos-ledger/bcos-ledger/Ledger.cpp:2158]: **⚡ 性能**: 当前实现对 `ethStateTrie.nodes` 中的每个节点执行一次 `co_await storage2::writeOne()`，这意味着 N 个节点就需要 N 次异步存储写入操作。对于有大量 alloc 的 genesis 配置，这可能产生显著的写放大。考虑：
1. 是否可以收集所有节点到一个 batch 后一次性写入？（如果 `storage2` 支持 batch write API）
2. 或者利用已有的 genesis batch 机制——当前 alloc flat rows 写入和节点写入是串行的，能否合并到同一个事务中？

不过对于 genesis 初始化（一次性操作），这个性能影响在实践中可接受。建议加 TODO 标注。

File: bcos-ledger/bcos-ledger/Ledger.cpp [hunk 7]
```diff
@@ -2142,6 +2146,29 @@ bool Ledger::buildGenesisBlock(
         co_await importGenesisState(
             genesis.m_allocs, *m_stateStorage, *m_blockFactory->cryptoSuite()->hashImpl());
 
+        // Scenario B (L2): block 1 builds the MPT incrementally on top of the genesis state
+        // root (buildAndCollect with the genesis root as parent) and reads the parent trie
+        // through "/mpt/" state rows, so every genesis trie node — account trie plus each
+        // account's storage sub-trie — must be persisted here, on the same storage the alloc
+        // flat rows above just went to. A missing node aborts block-1 execution loudly
```
Comment (morebtcg) [inline bcos-ledger/bcos-ledger/Ledger.cpp:2153]: **👍 功能设计**: `feature_l2_ethereum_compat` 门控逻辑正确——和 `NodeConfig::validateL2Invariants` 及 `shouldBuildMPT` 的场景 B 分支使用相同的谓词，保证了行为一致性。`std::any_of` 配合 `enable > 0` 检查也正确处理了 feature flag 的启用语义。

File: bcos-ledger/bcos-ledger/Ledger.cpp [hunk 8]
```diff
@@ -2142,6 +2146,29 @@ bool Ledger::buildGenesisBlock(
...
+        // (MPTInvariantViolation). Non-MPT chains never read these rows and get none; scenario
+        // A (feature_mpt_state_root activated mid-chain) starts its first MPT block from
+        // emptyRootHash() and needs no genesis nodes either.
```
Comment (morebtcg) [inline bcos-ledger/bcos-ledger/Ledger.cpp:2156]: **代码风格**: 写入循环中 `for (auto& [nodeHash, nodeRlp] : ethStateTrie.nodes)` 使用了 structured binding 和 `auto&`，并通过 `std::move(nodeRlp)` 避免了拷贝——良好的 C++17 实践。注意这里 move 后 `ethStateTrie.nodes` 中的元素处于 moved-from 状态，但后续不再使用，所以安全。

File: transaction-scheduler/tests/FullChainFixture.h [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+#include "bcos-ledger/mpt/StorageValueCodec.h"
+#include "bcos-protocol/TransactionSubmitResultFactoryImpl.h"
+#include "bcos-storage/CheckpointRocksDBStorage.h"
+#include "bcos-storage/KeyPrefixes.h"
+#include "bcos-storage/RocksDBStorage.h"
+#include "bcos-storage/StateKVResolver.h"
+#include "bcos-tars-protocol/protocol/BlockFactoryImpl.h"
+#include "bcos-tars-protocol/protocol/BlockHeaderFactoryImpl.h"
+#include "bcos-tars-protocol/protocol/BlockImpl.h"
+#include "bcos-tars-protocol/protocol/TransactionFactoryImpl.h"
```
Comment (morebtcg) [inline transaction-scheduler/tests/FullChainFixture.h:65]: **⚠️ 可扩展性/代码组织**: 这是一个 484 行的 header-only 测试 fixture，包含了模板类 (`FCWritingScheduler`, `FCExecutor`, `FCCheckpointStorage`, `FCMultiLayerStorage`)、完整的辅助函数和 RocksDB 初始化逻辑。问题在于：
1. 任何包含此头的 TU 都需要完整编译所有模板和 RocksDB 头文件
2. 多个测试文件包含它时（当前有 5 个），每个 TU 都要独立实例化 `BaselineScheduler<FCMultiLayerStorage, FCExecutor, ...>` 模板
3. 这就是为什么需要在 CMakeLists.txt 中为每个使用者设置 `SKIP_UNITY_BUILD_INCLUSION`

建议：将非模板部分（如 `TmpDirGuard`, `FullChainFixture` 的构造函数/大部分成员函数）提取到 `.cpp` 文件中，header 只保留模板和声明。这也会减少每个测试 TU 的编译时间。

File: transaction-scheduler/tests/FullChainFixture.h [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+#include "bcos-tars-protocol/protocol/TransactionReceiptFactoryImpl.h"
+#include "bcos-tars-protocol/protocol/TransactionReceiptImpl.h"
+#include "bcos-transaction-scheduler/BaselineScheduler.h"
+#include <boost/test/unit_test.hpp>
+#include <atomic>
+#include <filesystem>
+#include <map>
+#include <memory>
+#include <optional>
```
Comment (morebtcg) [inline transaction-scheduler/tests/FullChainFixture.h:74]: **👍 功能设计**: `TmpDirGuard` 的 RAII 设计值得称赞——声明为第一个成员确保析构顺序在 RocksDB handles 之后，`std::filesystem::remove_all` 在析构时忽略错误 (`ec` 参数)。`static atomic<uint64_t>` 计数器确保了并行测试的目录名唯一性。这些都是生产级测试基础设施的好实践。

File: transaction-scheduler/tests/FullChainFixture.h [hunk 11]
```diff
@@ -0,0 +1,484 @@
...
+ *        BaselineSchedulerMPTGenesisTest (test-transaction-scheduler) and
+ *        EthGetProofIntegrationTest (test-bcos-rpc).
+ */
+#pragma once
+
+#include "bcos-crypto/hash/Keccak256.h"
+#include "bcos-framework/ledger/Features.h"
+#include "bcos-framework/ledger/GenesisConfig.h"
+#include "bcos-framework/ledger/Ledger.h"
+#include "bcos-framework/ledger/LedgerTypeDef.h"
```
Comment (morebtcg) [inline transaction-scheduler/tests/FullChainFixture.h:45]: **⚠️ 性能/正确性**: `FCCheckpointStorage` 继承 `CheckpointRocksDBStorage`，其 `createDataTable` 返回的是栈上局部变量的指针，这是一个 **use-after-return** 的严重问题：
```cpp
auto createDataTable(std::string const&, storage::Entry) -> std::shared_ptr<rocksdb::Table>
{ return std::make_shared<rocksdb::Table>(); }
```
这里 `std::make_shared` 返回的是堆分配对象，实际上是安全的。但如果编译警告将此标记为潜在问题，建议明确标注为什么这里用 `shared_ptr` 是安全的。同时确认 `CheckpointRocksDBStorage` 基类对该返回值的生命周期管理是否正确。

File: bcos-rpc/test/CMakeLists.txt [hunk 12]
```diff
@@ -26,9 +26,21 @@ add_executable(${TEST_BINARY_NAME} ${SOURCES})
 target_include_directories(${TEST_BINARY_NAME} PRIVATE .)
 target_compile_options(${TEST_BINARY_NAME} PRIVATE -Wno-unused-variable)
 
-target_link_libraries(${TEST_BINARY_NAME} txpool ledger rpc tool Boost::unit_test_framework)
+# transaction-scheduler: EthGetProofIntegrationTest.cpp drives the FullChainFixture
+# end-to-end harness (transaction-scheduler/tests/FullChainFixture.h); it brings
+# ledger/storage (RocksDB) transitively. Deliberately NOT linking `executor` here even
+# though the BaselineScheduler instantiation references bcos::unhexAddress: executor
+# PUBLIC-links wedprcrypto's Rust static libraries, whose bundled runtime breaks libc++
```
Comment (morebtcg) [inline bcos-rpc/test/CMakeLists.txt:33]: **💡 可维护性**: 关于不链接 `executor` 的注释非常详尽——解释了 wedprcrypto Rust 静态库破坏 libc++ 异常捕获的问题。这类跨语言链接问题的文档化非常好，但建议将这个已知问题的根因记录到项目级别的 `docs/known-issues.md` 或仓库 memory 中，而不是只写在这个 CMakeLists.txt 的注释里，方便其他可能遇到同样问题的模块开发者参考。

**Review by morebtcg — COMMENTED** [review]
## Code Review: persist genesis trie nodes for L2 chains

整体设计思路清晰，将 genesis trie 节点从"只算根不存节点"改为"计算完整 trie 并持久化所有节点"，修复了 L2 链区块 1 的 `MPTInvariantViolation` 问题。以下是按文件细分的具体意见。

**总结评分：**
- 代码风格: ⭐⭐⭐⭐ (注释详尽，命名清晰)
- 功能设计: ⭐⭐⭐⭐ (正确解决了问题，边界处理周全)
- 可扩展性: ⭐⭐⭐ (header-only fixture 过重，`unordered_map` 迭代顺序需关注)
- 性能: ⭐⭐⭐ (逐条写入 vs 批量写入值得考量)

**Review by morebtcg — APPROVED** [review]
LGTM. All comments from the previous review are non-blocking suggestions. Core logic is correct, edge cases are well-handled, and test coverage is thorough. Approving.
